File: app/src/main/java/com/vidyarthi/bus/adapters/RouteAdapter.java
package com.vidyarthi.bus.adapters;

import android.graphics.Color;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.vidyarthi.bus.R;
import com.vidyarthi.bus.models.BusRoute;
import java.util.List;

public class RouteAdapter extends RecyclerView.Adapter<RouteAdapter.RouteViewHolder> {

    public interface OnRouteClickListener {
        void onRouteClick(BusRoute route);
    }

    private final List<BusRoute>        routes;
    private final OnRouteClickListener  listener;

    // Simulated crowd levels for the list view (in production, listen to Firebase)
    private static final int[]    CROWD_PERCENTS = { 28, 65, 96 };
    private static final String[] CROWD_LABELS   = { "Seats Free", "Getting Full", "Bus Full" };
    private static final int[]    CROWD_COLORS   = {
        Color.parseColor("#1AAD4E"),
        Color.parseColor("#F5A623"),
        Color.parseColor("#E8322A")
    };

    public RouteAdapter(List<BusRoute> routes, OnRouteClickListener listener) {
        this.routes   = routes;
        this.listener = listener;
    }

    @NonNull
    @Override
    public RouteViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext())
                               .inflate(R.layout.item_route, parent, false);
        return new RouteViewHolder(v);
    }

    @Override
    public void onBindViewHolder(@NonNull RouteViewHolder holder, int position) {
        BusRoute route = routes.get(position);
        holder.tvRouteName.setText(route.getRouteName());
        holder.tvRouteDesc.setText(route.getRouteDescription());

        // Determine crowd level colour for the status pill
        int level = position < 3 ? getDefaultLevel(position) : 0;
        holder.tvCrowdPill.setText(CROWD_LABELS[level]);
        holder.tvCrowdPill.setTextColor(CROWD_COLORS[level]);
        holder.tvCrowdPill.setBackgroundTintList(
            android.content.res.ColorStateList.valueOf(
                alphaColor(CROWD_COLORS[level], 30)));

        holder.itemView.setOnClickListener(v -> listener.onRouteClick(route));
    }

    @Override
    public int getItemCount() { return routes.size(); }

    private int getDefaultLevel(int position) {
        if (position == 0) return 0; // green
        if (position == 1) return 1; // amber
        return 2;                    // red
    }

    private int alphaColor(int color, int alpha) {
        return Color.argb(alpha,
                Color.red(color), Color.green(color), Color.blue(color));
    }

    static class RouteViewHolder extends RecyclerView.ViewHolder {
        TextView tvRouteName, tvRouteDesc, tvCrowdPill;

        RouteViewHolder(@NonNull View itemView) {
            super(itemView);
            tvRouteName  = itemView.findViewById(R.id.tv_route_name);
            tvRouteDesc  = itemView.findViewById(R.id.tv_route_desc);
            tvCrowdPill  = itemView.findViewById(R.id.tv_crowd_pill);
        }
    }
}

File: app/src/main/java/com/vidyarthibus/ui/home/RouteAdapter.kt
package com.vidyarthibus.ui.home

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.vidyarthibus.R
import com.vidyarthibus.utils.CrowdUtils

class RouteAdapter(private val onClick: (RouteUiItem) -> Unit) :
    ListAdapter<RouteUiItem, RouteAdapter.VH>(DIFF) {

    inner class VH(v: View) : RecyclerView.ViewHolder(v) {
        val tvName  : TextView = v.findViewById(R.id.tvRouteName)
        val tvStops : TextView = v.findViewById(R.id.tvRouteStops)
        val tvPill  : TextView = v.findViewById(R.id.tvCrowdPill)
        val tvIcon  : TextView = v.findViewById(R.id.tvRouteIcon)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
        VH(LayoutInflater.from(parent.context).inflate(R.layout.item_route, parent, false))

    override fun onBindViewHolder(h: VH, pos: Int) {
        val item = getItem(pos)
        val ctx  = h.itemView.context
        h.tvName.text  = "Route ${item.route.routeNumber}"
        h.tvStops.text = item.route.name
        h.tvIcon.text  = when (item.crowdState.level.ordinal) { 0 -> "🟢"; 1 -> "🟡"; else -> "🔴" }
        CrowdUtils.applyToPill(ctx, h.tvPill, item.crowdState)
        h.itemView.setOnClickListener { onClick(item) }
    }

    companion object {
        val DIFF = object : DiffUtil.ItemCallback<RouteUiItem>() {
            override fun areItemsTheSame(a: RouteUiItem, b: RouteUiItem) = a.route.id == b.route.id
            override fun areContentsTheSame(a: RouteUiItem, b: RouteUiItem) = a == b
        }
    }
}
